File: 1501~2000/1512. Number of Good Pairs/main.py
"""
Source: https://leetcode.com/problems/number-of-good-pairs/
Date: 2022/12/1
Skill: 
Runtime: 55 ms, faster than 67.24%
Memory Usage: 13.9 MB, less than 56.76%
Time complexity: 
Space complexity: 
Constraints: 
    1 <= nums.length <= 100
"""

from typing import List

class Solution:
    def numIdenticalPairs(self, nums: List[int]) -> int:
        res = 0
        mydict = dict()
        for n in nums:
            if n in mydict:
                res += mydict[n]
                mydict[n] += 1
            else:
                mydict[n] = 1
        return res

File: 1501~2000/1512. Number of Good Pairs/main.cpp
/**
 * Source: ibit.ly/RgEMP
 * Date: 2023/10/3
 * Skill:
 * Runtime: 0 ms, faster than 100.00% of C++ online submissions
 * Memory Usage: 7.47 MB, less than 27.67% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *      1 <= nums[i] <= 100
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>

using namespace std;

#define ll long long
typedef pair<int, int> pairs;
typedef unsigned long long ULL;
using PULL = pair<ULL, ULL>;


class Solution {
public:
    int numIdenticalPairs(vector<int>& nums) {
        int cnt[101] = {0}, res = 0;
        for (auto i: nums)
            cnt[i]++;
        for (int i = 1; i <= 100; i++)
            if (cnt[i] > 1)
                res += cnt[i] * (cnt[i] - 1) / 2;
        return res;
    }
};


int main() {
//    Solution s;
    vector<int> nums{2,3,3,2};
    vector<vector<int>> arr1{{0,1},{1,2},{1,3}};
}